rclcpp/src/rclcpp/executor.cpp

```diff
@@ -559,18 +412,13 @@ Executor::execute_any_executable(AnyExecutable & any_exec)
     execute_client(any_exec.client);
   }
   if (any_exec.waitable) {
-    any_exec.waitable->execute(any_exec.data);
+    auto data = any_exec.waitable->take_data();
```

Comment:
This change breaks the multi threaded executor.

take_data must be called, in a sequence of is_ready, take_data under the protection of a mutex.

Reply:
Hmm, I don't think that we ever state that explicitly anywhere.  Ideally, `is_ready` should be idempotent in that you can call it multiple times on the same waitset and it always returns true if the waitable is ready, false otherwise.

`take_data`, I believe, should reset the state of the waitable so that it no longer returns true to `is_ready`, and should be done immediately before executing the waitable.

Previously, I was calling `take_data` when getting the list of all of the executables out of the waitset, but this could cause a condition where `take_data` was called without ever executing on it.

Reply:
>  Ideally, is_ready should be idempotent in that you can call it multiple times on the same waitset and it always returns true if the waitable is ready, false otherwise.

This holds, and is not affected by where we call take_data

> take_data, I believe, should reset the state of the waitable so that it no longer returns true to is_ready, and should be done immediately before executing the waitable.

Calling directly before the execution would be fine. You just need to make sure that you are holding the mutex, that prevents someone else from calling is_ready at the same time. 

Currently you can run into this scenario:
- T1 is_ready is called, element added to list of ready executables
- T1 Mutex is released
- T2 Second thread waits, calls is_ready, and adds the element
- T1 Execution thread comes by, calls take_data, nothing ready after this
- T2 Calls take_data, but nothing is there.

Reply:
This results in a wasted call to `take_data`, but to be clear it's ok to call `take_data` when nothing is ready, it's just not optimal and so it should be avoided when possible.

There are other reasons that something can happen between `is_ready` and `take_data` which will result in `take_data` having nothing to return. For example, if a subscription is set up with the lifespan QoS then `is_ready` could return true, then the message that make it ready expires and is removed, then `take_data` is called but the underlying DDS tells it that there is nothing to take. And that scenario is totally fine, but a bit of a waste of resources, which should be avoided if possible but isn't a bug or anything (in my opinion).

The multi-threaded issue you're describing above should probably be avoided, if possible, but I don't think it breaks anything exactly.

I could see that there should be a `if data == nullptr` check that avoids calling execute if there was no data to take, avoiding a similar check in every execute implementation.

Reply:
> This results in a wasted call to `take_data`, but to be clear it's ok to call `take_data` when nothing is ready, it's just not optimal and so it should be avoided when possible.
> 
Doing this will crash the you executable if using actions:
https://github.com/ros2/rclcpp/blob/99f0fc938b2d0c441787113a3920c16d3614a197/rclcpp_action/src/client.cpp#L592

Reply:
Discussed this with @wjwwood this afternoon and realized that this change was actually an artifact of the "ready executables queue" that was added.  I'm now removing that so this change will be un-done.